homeassistant/components/android_ip_webcam.py

```diff
@@ -0,0 +1,317 @@
+            'motion_detect', 'night_vision', 'overlay', 'torch',
+            'whitebalance_lock']
+
+SENSORS = ['audio_connections', 'battery_level', 'battery_temp',
+           'battery_voltage', 'light', 'motion', 'pressure', 'proximity',
+           'sound', 'video_connections']
+
+CONF_MOTION_BINARY_SENSOR = 'motion_binary_sensor'
+
+DEFAULT_MOTION_BINARY_SENSOR = True
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Required(CONF_HOST): cv.string,
+        vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+        vol.Inclusive(CONF_USERNAME, 'authentication'): cv.string,
+        vol.Inclusive(CONF_PASSWORD, 'authentication'): cv.string,
+        vol.Optional(CONF_SWITCHES,
+                     default=SWITCHES): vol.All(cv.ensure_list,
+                                                [vol.In(SWITCHES)]),
+        vol.Optional(CONF_SENSORS,
+                     default=SENSORS): vol.All(cv.ensure_list,
+                                               [vol.In(SENSORS)]),
+        vol.Optional(CONF_MOTION_BINARY_SENSOR,
+                     default=DEFAULT_MOTION_BINARY_SENSOR): bool
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+ALLOWED_ORIENTATIONS = ['landscape', 'upsidedown', 'portrait',
+                        'upsidedown_portrait']
+
+MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=3)
+
+
+def setup(hass, config):
+    """Setup the IP Webcam component."""
+    conf = config[DOMAIN]
+    host = conf[CONF_HOST]
+    ip_webcam = hass.data.get(DATA_IP_WEBCAM)
+    if ip_webcam is None:
+        hass.data[DATA_IP_WEBCAM] = {}
+    hass.data[DATA_IP_WEBCAM][host] = IPWebcam(conf)
+
+    if conf.get(CONF_MOTION_BINARY_SENSOR, False) is True:
+        discovery.load_platform(hass, 'binary_sensor', DOMAIN, {}, config)
+
+    discovery.load_platform(hass, 'camera', DOMAIN, {}, config)
+
+    sensor_config = conf.get(CONF_SENSORS, [])
+    discovery.load_platform(hass, 'sensor', DOMAIN, sensor_config, config)
+
+    switch_config = conf.get(CONF_SWITCHES, [])
+    discovery.load_platform(hass, 'switch', DOMAIN, switch_config, config)
+
+    return True
+
+
+class IPWebcam(object):
+    """The Android device running IP Webcam."""
+
+    def __init__(self, config):
+        """Initialize the data oject."""
+        self._config = config
+        self._name = self._config.get(CONF_NAME)
+        self.host = self._config.get(CONF_HOST)
+        self.port = self._config.get(CONF_PORT)
+        self.username = self._config.get(CONF_USERNAME)
+        self.password = self._config.get(CONF_PASSWORD)
+        self.status_data = None
+        self.sensor_data = None
+        self._sensor_updated_at = (datetime.now() - timedelta(seconds=5))
+        self.update()
+
+    @property
+    def base_url(self):
+        """Return the base url for endpoints."""
+        return 'http://{}:{}'.format(self.host, self.port)
+
+    def _request(self, path):
```

Comment:
When you extract it into a third party lib, have it take an aiohttp session in the constructor. That way you don't need to do any closing work because you can use the one managed by hass.